java/src/main/native/include/jni_utils.hpp

```diff
@@ -862,6 +862,15 @@ inline void jni_cuda_check(JNIEnv *const env, cudaError_t cuda_status) {
     JNI_CHECK_CUDA_ERROR(env, cudf::jni::CUDA_ERROR_CLASS, e, ret_val);                            \
   }                                                                                                \
   catch (const std::exception &e) {                                                                \
+    /* Double check whether the thrown exception is unrecoverable CUDA error or not. */            \
+    /* Like cudf::detail::throw_cuda_error, it is nearly certain that a fatal error  */            \
+    /* occurred if the second call doesn't return with cudaSuccess. */                             \
+    auto const last = cudaDeviceSynchronize();                                                     \
+    if (cudaSuccess != last && last == cudaGetLastError()) {                                       \
```

Comment:
Do we need an extra `cudaGetLastError()` here as in the original libcudf code?  Detecting a fatal error seems more about checking if `cudaGetLastError` keeps returning the same error.